io/StacReader.cpp

```diff
@@ -127,112 +123,206 @@ void StacReader::addArgs(ProgramArgs& args)
+
+    Stage *stage = m_factory.createStage(driver);
+
+    if (!stage)
+    {
+        std::stringstream msg;
+        msg << "Unable to create driver '" << driver << "' for "
+            << "for asset located at '" << item.assetPath() <<"'";
+        throw pdal_error(msg.str());
+    }
+    Reader* reader = dynamic_cast<Reader*>(stage);
+    if (!reader)
+    {
+        std::stringstream msg;
+        msg << "Unable to cast stage of type '" << driver << "' to a reader";
+        throw pdal_error(msg.str());
+    }
+    reader->setOptions(item.options());
+
+    std::lock_guard<std::mutex> lock(m_p->m_mutex);
+    m_p->m_idList.push_back(item.id());
+    reader->setLog(log());
+    m_merge.setInput(*reader);
+    m_p->m_readerList.push_back(reader);
+}
 
+void StacReader::handleItem(NL::json stacJson, std::string itemPath)
+{
+    stac::Item item(stacJson, m_filename, *m_p->m_connector,
+        m_args->validateSchema);
+    if (item.init(*m_p->m_itemFilters, m_args->rawReaderArgs, m_args->schemaUrls))
+        addItem(item);
 }
 
 
-void StacReader::validateSchema(NL::json stacJson)
+void StacReader::handleCatalog(NL::json stacJson, std::string catPath, bool isRoot)
 {
-    std::function<void(const nlohmann::json_uri&, nlohmann::json&)> fetch = schemaFetch;
-    nlohmann::json_schema::json_validator val(
-        fetch,
-        [](const std::string &, const std::string &) {}
-    );
-    if (!stacJson.contains("type"))
-        throw pdal_error("Invalid STAC json");
-    std::string type = stacJson.at("type").get<std::string>();
-    std::string schemaUrl;
-
-    if (type == "Feature")
+    stac::Catalog catalog(stacJson, catPath, *m_p->m_connector, *m_p->m_pool,
+        m_args->validateSchema);
+
+    if (catalog.init(*m_p->m_catFilters, m_args->rawReaderArgs,
+        m_args->schemaUrls, true))
     {
-        schemaUrl = m_args->featureSchemaUrl;
-        for (auto& extSchemaUrl: stacJson.at("stac_extensions"))
+        for (stac::Item& item: catalog.items())
+            addItem(item);
+    }
+    stac::ErrorList errors = catalog.errors();
+    if (errors.size())
+    {
+        for (auto& p: errors)
         {
-            log()->get(LogLevel::Debug) << "Processing extension " << extSchemaUrl << std::endl;
-            NL::json schemaJson = m_p->m_connector->getJson(extSchemaUrl);
-            val.set_root_schema(schemaJson);
-            val.validate(stacJson);
+            log()->get(LogLevel::Error) << "Failure fetching '" << p.first
+                << "' with error '" << p.second << "'";
         }
     }
-    else if (type == "Catalog")
-        schemaUrl = m_args->catalogSchemaUrl;
-    else
-        throw pdal_error("Invalid STAC type for PDAL consumption");
-
-    NL::json schemaJson = m_p->m_connector->getJson(schemaUrl);
-    val.set_root_schema(schemaJson);
-    val.validate(stacJson);
 }
 
-void StacReader::handleReaderArgs()
+void StacReader::handleCollection(NL::json stacJson, std::string catPath, bool isRoot)
```

Comment:
I assume this one should be `colPath` instead of `catPath` which was presumably copy/pasted from the `catalog` version.  Same is true in the header prototypes.  Seems reasonable to just call them `path` or `filename` in this context since they are scoped to a function whose name already describes them.